services/src/main/java/org/exoplatform/addons/gamification/listener/challenges/ChallengesESListener.java

```diff
@@ -0,0 +1,74 @@
+package org.exoplatform.addons.gamification.listener.challenges;
+
+import org.exoplatform.addons.gamification.connector.ChallengesIndexingServiceConnector;
+import org.exoplatform.addons.gamification.service.ChallengeService;
+import org.exoplatform.addons.gamification.service.dto.configuration.Challenge;
+import org.exoplatform.addons.gamification.utils.Utils;
+import org.exoplatform.commons.search.index.IndexingService;
+import org.exoplatform.container.ExoContainerContext;
+import org.exoplatform.container.PortalContainer;
+import org.exoplatform.container.component.RequestLifeCycle;
+import org.exoplatform.services.listener.Event;
+import org.exoplatform.services.listener.Listener;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+
+public class ChallengesESListener extends Listener<Long, Object> {
+  private static final Log      LOG = ExoLogger.getLogger(ChallengesESListener.class);
+
+  private final PortalContainer container;
+
+  private final IndexingService indexingService;
+
+  private ChallengeService      challengeService;
+
+  public ChallengesESListener(PortalContainer container, IndexingService indexingService) {
+    this.container = container;
+    this.indexingService = indexingService;
+  }
+
+  @Override
+  public void onEvent(Event<Long, Object> event) throws Exception {
+    ExoContainerContext.setCurrentContainer(container);
+    RequestLifeCycle.begin(container);
+
+    Long challengeId = event.getSource();
+    try {
+      if (indexingService != null) {
+        if (!Utils.POST_DELETE_CHALLENGE_EVENT.equals(event.getEventName())) {
+          Challenge challenge = getChallengeService().getChallengeById(challengeId, Utils.getCurrentUser());
```

Comment:
No need to get chaqllengeService from the Container, we may pass it as the source of the event